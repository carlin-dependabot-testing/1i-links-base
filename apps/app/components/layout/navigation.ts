import {
  BarChart3,
  BookMarked,
  Clock,
  Home,
  LayoutDashboard,
  type LucideIcon,
  Scroll,
  Settings,
  Tags,
  Users
} from 'lucide-react'

import appConfig from '@/config/app.config'
import { ROUTES } from '@/constants/routes'

export type NavItem = {
  title: string
  url: string
  icon: LucideIcon
  pageTitle: string
  description: string
}

export const ADDITIONAL_NAV_ITEMS = [
  {
    title: 'Recently Added',
    url: ROUTES.RECENT.path,
    icon: Clock,
    pageTitle: 'Recently Added Links',
    description: 'View the most recently added links in your collection'
  },
  {
    title: 'Analytics',
    url: ROUTES.ANALYTICS.path,
    icon: BarChart3,
    pageTitle: 'Analytics Dashboard',
    description: 'View insights and statistics about your links'
  }
] as const

export type AdminNavItem = {
  title: string
  url: string
  icon: LucideIcon
}

export const ADMIN_NAVIGATION: { navMain: AdminNavItem[] } = {
  navMain: [
    {
      title: 'Dashboard',
      url: '/admin',
      icon: LayoutDashboard
    },
    {
      title: 'Manage Links',
      url: '/admin/manage/links',
      icon: BookMarked
    },
    {
      title: 'Manage Categories',
      url: '/admin/manage/categories',
      icon: Users
    },
    {
      title: 'Manage Tags',
      url: '/admin/manage/tags',
      icon: Tags
    },
    {
      title: 'Settings',
      url: '/admin/settings',
      icon: Settings
    }
  ]
} as const

export const ADMIN_NAVIGATION_SUPPORT = [
  {
    title: 'Changelog',
    url: '/admin/changelog',
    icon: Scroll
  }
] as const

export type AdminEnvironment = {
  name: string
  path: string
}

export const ADMIN_ENVIRONMENTS: AdminEnvironment[] = [
  {
    name: appConfig.name,
    path: `${appConfig.url}`
  },
  {
    name: 'Admin Panel',
    path: `${appConfig.url}/admin`
  }
]

export const MAIN_NAVIGATION = [
  {
    title: 'Home',
    url: `${appConfig.url}`,
    icon: Home,
    isActive: true
  }
] as const
