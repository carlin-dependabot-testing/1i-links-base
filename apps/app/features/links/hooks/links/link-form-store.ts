import { create } from 'zustand'
import { persist } from 'zustand/middleware'

import { DEFAULT_LINK_VALUES } from '@/features/links/constants/link-form'
import { type LinksApp } from '@/features/links/types/link.types'

interface LinkFormState {
  formData: Partial<LinksApp>
  mode: 'add' | 'edit'
  originalName: string | null
  setFormData: (data: Partial<LinksApp>) => void
  setMode: (mode: 'add' | 'edit') => void
  setOriginalName: (name: string | null) => void
  resetForm: () => void
  updateField: (field: string, value: any) => void
}

export const useLinkFormStore = create<LinkFormState>()(
  persist(
    set => ({
      formData: DEFAULT_LINK_VALUES,
      mode: 'add',
      originalName: null,
      setFormData: data => {
        set(() => ({
          formData: {
            ...DEFAULT_LINK_VALUES,
            ...data,
            environments: {
              production: data.environments?.production || '',
              staging: data.environments?.staging || '',
              integration: data.environments?.integration || ''
            },
            tags: data.tags || []
          }
        }))
      },
      setMode: mode => set({ mode }),
      setOriginalName: name => set({ originalName: name }),
      resetForm: () =>
        set({
          formData: DEFAULT_LINK_VALUES,
          mode: 'add',
          originalName: null
        }),
      updateField: (field: string, value: any) => {
        set(state => {
          const newState = { ...state.formData }

          // Handle nested fields like 'environments.staging'
          if (field.includes('.')) {
            const [parent, child] = field.split('.') as [keyof LinksApp, string]
            if (!newState[parent]) {
              newState[parent] =
                parent === 'environments'
                  ? { production: '', staging: '', integration: '' }
                  : parent === 'tags'
                    ? ([] as string[])
                    : ({} as any)
            }
            if (typeof newState[parent] === 'object') {
              ;(newState[parent] as Record<string, any>)[child] = value
            }
          } else {
            newState[field as keyof LinksApp] = value
          }

          return { formData: newState }
        })
      }
    }),
    {
      name: 'link-form-storage'
    }
  )
)
