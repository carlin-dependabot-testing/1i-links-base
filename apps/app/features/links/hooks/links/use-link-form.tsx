import { useForm } from 'react-hook-form'

import { zodResolver } from '@hookform/resolvers/zod'

import {
  type InferLinkSchema,
  LinkSchema
} from '@/features/links/schemas/link.schema'

export const useLinkForm = (
  initialValues: Partial<InferLinkSchema> = {},
  onSubmit: (data: InferLinkSchema) => Promise<void>
) => {
  const form = useForm<InferLinkSchema>({
    resolver: zodResolver(LinkSchema),
    defaultValues: {
      name: initialValues.name ?? '',
      environments: initialValues.environments ?? { production: '' },
      category: initialValues.category ?? '',
      tags: initialValues.tags ?? []
    }
  })

  const handleSubmit = async (data: InferLinkSchema) => {
    try {
      await onSubmit(data)
      form.reset()
      return true
    } catch (error) {
      console.error('Failed to submit link:', error)
      return false
    }
  }

  return {
    form,
    handleSubmit: form.handleSubmit(handleSubmit),
    isSubmitting: form.formState.isSubmitting
  }
}
