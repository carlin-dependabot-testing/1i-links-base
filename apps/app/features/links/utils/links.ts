import { DEFAULT_CATEGORY } from '@/constants'
import { type LinksApp } from '@/features/links/types/link.types'

export interface GroupLinksByCategoryParams {
  links: LinksApp[]
  category?: string
  isFavourite: (name: string) => boolean
  isArchived: (name: string) => boolean
  showArchived: boolean
  hiddenCategories: string[]
  validCategories: string[] | Set<string>
}

export function groupLinksByCategory({
  links,
  category,
  isFavourite,
  isArchived,
  showArchived,
  hiddenCategories,
  validCategories
}: GroupLinksByCategoryParams): Record<string, LinksApp[]> {
  const categoriesSet =
    validCategories instanceof Set
      ? validCategories
      : new Set((validCategories || []).map(cat => cat.toLowerCase()))

  return links
    .filter(link => !isFavourite(link.name))
    .filter(link => (showArchived ? true : !isArchived(link.name)))
    .filter(link => link.category && !hiddenCategories.includes(link.category))
    .reduce(
      (acc, link) => {
        const linkCategory = link.category || DEFAULT_CATEGORY

        // Case-insensitive category comparison
        if (category && category.toLowerCase() !== linkCategory.toLowerCase()) {
          return acc
        }

        if (categoriesSet.has(linkCategory.toLowerCase())) {
          if (!acc[linkCategory]) {
            acc[linkCategory] = []
          }
          acc[linkCategory].push(link)
        }

        return acc
      },
      {} as Record<string, LinksApp[]>
    )
}
