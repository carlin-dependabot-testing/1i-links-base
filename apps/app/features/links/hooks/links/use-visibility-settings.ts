'use client'

import { useEffect, useState } from 'react'

import { useSearchParams } from 'next/navigation'

import { useLocalStorage } from 'usehooks-ts'

import { STORAGE_KEYS } from '@/constants/storage'

export const useVisibilitySettings = () => {
  const [localShowArchived, setLocalShowArchived] = useLocalStorage(
    STORAGE_KEYS.SHOW_ARCHIVED,
    false
  )
  const [showArchived, setShowArchived] = useState(false)
  const searchParams = useSearchParams()

  // URL parameter takes precedence over localStorage
  const urlShowArchived = searchParams.get('archived') === 'true'

  // Only update on mount and when URL/localStorage changes
  useEffect(() => {
    const shouldShow = urlShowArchived ?? localShowArchived
    if (showArchived !== shouldShow) {
      setShowArchived(shouldShow)
    }
  }, [urlShowArchived, localShowArchived])

  const updateShowArchived = (value: boolean) => {
    setLocalShowArchived(value)
    setShowArchived(value)
  }

  return {
    showArchived,
    setShowArchived: updateShowArchived
  }
}
