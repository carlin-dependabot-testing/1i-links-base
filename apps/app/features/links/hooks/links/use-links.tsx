import { useEffect, useState } from 'react'

import { useRouter } from 'next/navigation'

import { getLinkApiRoute, getLinksPath } from '@/constants/routes'
import { type InferLinkSchema } from '@/features/links/schemas/link.schema'
import { type LinksApp } from '@/features/links/types/link.types'
import { useSetupStore } from '@/features/setup/stores/setupStore'

import { useToast } from '@links-base/ui/hooks'
import { ToastAction } from '@links-base/ui/toast'

/**
 * Hook to manage links
 * @returns {Object} - Returns an object with the links, isLoading, error, refreshLinks, addLink, deleteLink, and editLink functions
 */
export function useLinks() {
  const [links, setLinks] = useState<LinksApp[]>([])
  const [isLoading, setLoading] = useState(true)
  const [error, setError] = useState<string | null>(null)
  const { toast } = useToast()
  const router = useRouter()
  const { isComplete, resetSetup } = useSetupStore()

  const fetchLinks = async () => {
    setLoading(true)
    setError(null)
    try {
      const response = await fetch(getLinksPath(), {
        cache: 'no-store',
        headers: {
          'Cache-Control': 'no-cache'
        }
      })

      if (!response.ok) {
        // Check if the error is due to missing files
        if (response.status === 404) {
          resetSetup()
          router.push('/setup')
          return
        }
        throw new Error('Failed to fetch links')
      }

      const data = await response.json()
      setLinks(data || [])
    } catch (error) {
      const errorMessage = 'There was an error loading the links.'
      setError(errorMessage)
      console.error('Error fetching links:', error)

      // Only show error toast if setup is complete
      if (isComplete) {
        toast({
          title: 'Failed to load links',
          description: errorMessage,
          variant: 'destructive'
        })
      }
    } finally {
      setLoading(false)
    }
  }

  /**
   * Add a new link
   * @param link - The link object to add
   * @returns {Promise<boolean>} - Returns true if the link was added successfully
   */
  const addLink = async (link: InferLinkSchema) => {
    try {
      const response = await fetch(getLinkApiRoute(), {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(link)
      })

      if (!response.ok) throw new Error('Failed to add link')

      await fetchLinks() // Refresh links after adding
      toast({
        title: 'Link added successfully',
        description: 'The link has been added to the database.'
      })
      return true
    } catch (error) {
      console.error('Error adding link:', error)
      toast({
        title: 'Failed to add link',
        description: 'There was an error adding the link to the database.',
        variant: 'destructive'
      })
      return false
    }
  }

  /**
   * Delete a link with undo capability
   * @param linkName - The name of the link to delete
   * @returns {Promise<boolean>} - Returns true if the link was deleted successfully
   */
  const deleteLink = async (linkName: string) => {
    const linkToDelete = links.find(link => link.name === linkName)

    try {
      const response = await fetch(getLinkApiRoute(), {
        method: 'DELETE',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ linkName })
      })

      if (!response.ok) throw new Error('Failed to delete link')

      await fetchLinks() // Refresh links after deleting
      toast({
        title: 'Link deleted successfully',
        description: 'The link has been removed from the database.',
        action: linkToDelete ? (
          <ToastAction
            altText="Undo link deletion"
            onClick={() => {
              void addLink(linkToDelete).catch(error => {
                console.error('Error restoring link:', error)
                toast({
                  title: 'Failed to restore link',
                  description: 'There was an error restoring the link.',
                  variant: 'destructive'
                })
              })
            }}
          >
            Undo
          </ToastAction>
        ) : undefined
      })
      return true
    } catch (error) {
      toast({
        title: 'Failed to delete link',
        description: 'There was an error removing the link from the database.',
        variant: 'destructive'
      })
      return false
    }
  }

  /**
   * Edit a link
   * @param linkName - The name of the link to edit
   * @param updatedLink - The updated link data
   * @returns {Promise<boolean>} - Returns true if the link was edited successfully
   */
  const editLink = async (
    linkName: string,
    updatedLink: Partial<InferLinkSchema>
  ) => {
    try {
      const response = await fetch(getLinkApiRoute(), {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          linkName,
          ...updatedLink
        })
      })

      if (!response.ok) throw new Error('Failed to edit link')

      await fetchLinks() // Refresh links after editing
      toast({
        title: 'Link updated successfully',
        description: 'The link has been updated in the database.'
      })
      return true
    } catch (error) {
      console.error('Error editing link:', error)
      toast({
        title: 'Failed to update link',
        description: 'There was an error updating the link in the database.',
        variant: 'destructive'
      })
      return false
    }
  }

  useEffect(() => {
    void fetchLinks()
  }, []) // Only fetch on mount

  return {
    links,
    isLoading,
    error,
    refreshLinks: fetchLinks,
    addLink,
    deleteLink,
    editLink
  }
}
