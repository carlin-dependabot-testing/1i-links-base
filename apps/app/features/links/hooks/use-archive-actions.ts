'use client'

import { usePathname, useRouter, useSearchParams } from 'next/navigation'

import { useVisibilitySettings } from './links/use-visibility-settings'

import { useArchivedLinks } from './use-archived-links'

export const useArchiveActions = () => {
  const searchParams = useSearchParams()
  const pathname = usePathname()
  const router = useRouter()
  const { showArchived, setShowArchived } = useVisibilitySettings()
  const { toggleArchived, isArchived } = useArchivedLinks()

  const updateUrlParams = (updates: Record<string, string | null>) => {
    const params = new URLSearchParams(searchParams.toString())
    Object.entries(updates).forEach(([key, value]) => {
      if (value === null) {
        params.delete(key)
      } else {
        params.set(key, value)
      }
    })
    router.push(`${pathname}?${params.toString()}`)
  }

  const handleToggleArchived = (linkName: string) => {
    toggleArchived(linkName)
    // When archiving a new link, hide archived links
    if (!isArchived(linkName)) {
      setShowArchived(false)
      updateUrlParams({ archived: 'false' })
    }
  }

  return {
    showArchived,
    isArchived,
    handleToggleArchived
  }
}
