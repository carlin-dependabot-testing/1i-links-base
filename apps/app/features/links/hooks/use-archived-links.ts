import { usePathname, useRouter, useSearchParams } from 'next/navigation'

import { useLocalStorage } from 'usehooks-ts'

import { toast } from '@links-base/ui/hooks'

import { useVisibilitySettings } from './links/use-visibility-settings'

export const useArchivedLinks = () => {
  const [archivedLinks, setArchivedLinks] = useLocalStorage<string[]>(
    'archivedLinks',
    []
  )
  const { setShowArchived } = useVisibilitySettings()
  const searchParams = useSearchParams()
  const pathname = usePathname()
  const router = useRouter()

  const toggleArchived = (linkName: string) => {
    setArchivedLinks(prev => {
      const isArchiving = !prev.includes(linkName)
      const newArchivedLinks = isArchiving
        ? [...prev, linkName]
        : prev.filter(name => name !== linkName)

      // If we're archiving a new link, always ensure archived links are hidden
      if (isArchiving) {
        setShowArchived(false)
        // Update URL to hide archived links
        const params = new URLSearchParams(searchParams.toString())
        params.delete('archived')
        router.push(`${pathname}?${params.toString()}`, { scroll: false })
      }

      return newArchivedLinks
    })
  }

  const isArchived = (linkName: string) => {
    return archivedLinks.includes(linkName)
  }

  const unarchiveAll = () => {
    setArchivedLinks([])

    // Hide archived view
    setShowArchived(false)

    // Update URL
    const params = new URLSearchParams(searchParams.toString())
    params.delete('archived')
    params.delete('archivedLinks')
    router.push(`${pathname}?${params.toString()}`, { scroll: false })

    // Show toast
    toast({
      title: 'All links unarchived',
      description: `${archivedLinks.length} links have been restored`
    })
  }

  const getArchivedLinks = () => {
    return archivedLinks
  }

  return {
    archivedLinks,
    toggleArchived,
    isArchived,
    unarchiveAll,
    getArchivedLinks
  }
}
