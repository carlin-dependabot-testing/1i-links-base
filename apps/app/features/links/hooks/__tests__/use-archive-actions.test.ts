import { usePathname, useRouter, useSearchParams } from 'next/navigation'

import { act, renderHook } from '@testing-library/react'

import { useVisibilitySettings } from '../links/use-visibility-settings'
import { useArchiveActions } from '../use-archive-actions'
import { useArchivedLinks } from '../use-archived-links'

// Mock the Next.js navigation hooks
jest.mock('next/navigation', () => ({
  useSearchParams: jest.fn(),
  useRouter: jest.fn(),
  usePathname: jest.fn()
}))

// Mock the custom hooks
jest.mock('../use-archived-links', () => ({
  useArchivedLinks: jest.fn()
}))

jest.mock('../links/use-visibility-settings', () => ({
  useVisibilitySettings: jest.fn()
}))

describe('useArchiveActions', () => {
  const mockPush = jest.fn()
  const mockSetShowArchived = jest.fn()
  const mockToggleArchived = jest.fn()
  const mockIsArchived = jest.fn()

  beforeEach(() => {
    jest.clearAllMocks()

    // Setup mock implementations
    ;(useSearchParams as jest.Mock).mockReturnValue({
      toString: () => ''
    })
    ;(useRouter as jest.Mock).mockReturnValue({ push: mockPush })
    ;(usePathname as jest.Mock).mockReturnValue('/test-path')
    ;(useVisibilitySettings as jest.Mock).mockReturnValue({
      showArchived: false,
      setShowArchived: mockSetShowArchived
    })
    ;(useArchivedLinks as jest.Mock).mockReturnValue({
      toggleArchived: mockToggleArchived,
      isArchived: mockIsArchived
    })
  })

  it('useArchiveActions should render correctly', () => {
    const { result } = renderHook(() => useArchiveActions())
    expect(result.current).toMatchSnapshot()
  })

  it('should handle toggling archive status when archiving a link', () => {
    mockIsArchived.mockReturnValue(false)
    const { result } = renderHook(() => useArchiveActions())

    act(() => {
      result.current.handleToggleArchived('test-link')
    })

    expect(mockToggleArchived).toHaveBeenCalledWith('test-link')
    expect(mockSetShowArchived).toHaveBeenCalledWith(false)
    expect(mockPush).toHaveBeenCalledWith('/test-path?archived=false')
  })

  it('should handle toggling archive status when unarchiving a link', () => {
    mockIsArchived.mockReturnValue(true)
    const { result } = renderHook(() => useArchiveActions())

    act(() => {
      result.current.handleToggleArchived('test-link')
    })

    expect(mockToggleArchived).toHaveBeenCalledWith('test-link')
    expect(mockSetShowArchived).not.toHaveBeenCalled()
    expect(mockPush).not.toHaveBeenCalled()
  })

  it('should expose the correct properties', () => {
    const { result } = renderHook(() => useArchiveActions())

    expect(result.current).toHaveProperty('showArchived')
    expect(result.current).toHaveProperty('isArchived')
    expect(result.current).toHaveProperty('handleToggleArchived')
  })
})
