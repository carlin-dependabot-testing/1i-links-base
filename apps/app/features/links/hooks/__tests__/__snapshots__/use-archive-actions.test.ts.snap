// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useArchiveActions useArchiveActions should render correctly 1`] = `
{
  "handleToggleArchived": [Function],
  "isArchived": [MockFunction],
  "showArchived": false,
}
`;
